src/metabase/models/card.clj

```diff
@@ -840,6 +855,64 @@
     ;; Some sort of odd query like an aggregation over cards. Just return it as-is.
     card))
 
+(defonce ^{:doc "Set of card IDs which have been read, which are in `:not-started` state.
+           These cards should be prioritized for analysis."}
+  cards-for-priority-analysis
+  (atom #{}))
+
+(def ^:dynamic *upstream-cards-without-idents*
+  "Set of card IDs which were read during analysis of another card and which did not have valid idents
+  already. In other words, this is the set of cards which are blocking a :blocked card."
+  nil)
+
+(defn- mark-card-used-for-priority-analysis
+  "If we selected `:metadata_analysis_state` and it's `:not-started` then add this card's ID to the set of cards
+  we want to prioritize for analysis."
+  [{state :metadata_analysis_state :as card}]
+  ;; Mark :not-started cards as eligible for :priority analysis.
+  (when (= state :not-started)
+    (swap! cards-for-priority-analysis conj (:id card)))
+  ;; When backfilling a card (indicated by `*upstream-cards-without-idents*`), if we see a card which is not :executed
+  ;; or :analyzed it's a potential blocker - put it in the list of blockers.
+  (when (and *upstream-cards-without-idents*
+             (not (#{:executed :analyzed} (:metadata_analysis_state card))))
+    (swap! *upstream-cards-without-idents* conj (:id card)))
+
+  ;; Always returning the card unchanged.
+  card)
+
+(defn- has-valid-ident?
+  "Checks that this column both has an `:ident` and that it's valid.
+
+  For a native card, the ident must be based on this card's `entity_id`.
+
+  For a model, the ident must likewise be for this card, **and** `:model/inner_ident` must also be set to the
+  corresponding unwrapped, inner ident."
+  [{:keys [ident model/inner_ident] :as _column} {:keys [entity_id] :as card}]
+  (if (= (:type card) :model)
+    (and ident
+         (lib/valid-model-ident? ident entity_id)
+         inner_ident
+         (= ident (lib/model-ident inner_ident entity_id)))
+    (and ident (lib/valid-basic-ident? ident entity_id))))
```

Comment:
Just a nit, but I think sharing the non-model checks is more readable, and makes the distinction a bit more obvious.

```clojure
(and ident (lib/valid-basic-ident? ident entity_id)
     (or (not= :model (:type card))
         (and inner_ident (= ident (lib/model-ident inner_ident entity_id)))))
```

Reply:
Also what do you think about introducing a nice semantic predicate here instead of checking the type directly? 😉

Reply:
I approve of the general principle but it's a bit unexpected here. If I were to do that it would come with a refactor of other usage of `:type` in the BE and lib, which is out of scope here.

Especially when it's not used widely to start with, the unfortunate overloading of the word "model" in a Toucan namespace makes it confusing.

IMO, a tech debt issue for another day.

Reply:
Do you disagree about regrouping the booleans? Just checking that suggestion wasn't missed because of the predicate comment.

I don't see a problem in slowly introducing the predicates, especially if it means avoiding adding more naked :type checks. Maybe I'm misunderstanding how you want to do the encapsulation? I was thinking of one like this for now:

```
(defn has-model-ident? [card] 
  (= :model (:type card)))
```